edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,275 @@
+                .ToList();
+
+            foreach (ServiceIdentityTreeNode child in danglingChildren)
+            {
+                child.Parent = Option.Some(newNode);
+                newNode.Children.Add(child);
+            }
+        }
+
+        void RemoveSingleNode_NoCallback(IDictionary<string, ServiceIdentityTreeNode> nodes, string id)
+        {
+            if (nodes.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from the parent
+                target.Parent.ForEach(parent => parent.Children.Remove(target));
+
+                // Unhook the children
+                target.Children.ForEach(child => child.Parent = Option.None<ServiceIdentityTreeNode>());
+
+                // Remove
+                nodes.Remove(id);
+            }
+        }
+
+        void RemoveRecursive(IDictionary<string, ServiceIdentityTreeNode> nodes, string id)
+        {
+            if (nodes.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from the parent
+                target.Parent.ForEach(parent => parent.Children.Remove(target));
+
+                // Recursively remove each child
+                var childrenSnapshot = new List<ServiceIdentityTreeNode>(target.Children);
+                childrenSnapshot.ForEach(child => this.RemoveRecursive(nodes, child.Identity.Id));
+
+                // Remove self after recursion
+                nodes.Remove(id);
+
+                // Notify callbacks
+                this.ServiceIdentityRemoved?.Invoke(this, id);
+            }
+        }
+
+        Option<ServiceIdentityTreeNode> FindDeviceByScopeId(IDictionary<string, ServiceIdentityTreeNode> nodes, string scopeId)
+        {
+            Preconditions.CheckNonWhiteSpace(scopeId, nameof(scopeId));
+
+            // Look for Edge devices with a matching scope ID
+            List<ServiceIdentityTreeNode> devices =
+                nodes
+                .Select(kvp => kvp.Value)
+                .Where(serviceIdentity => serviceIdentity.Identity.IsEdgeDevice && serviceIdentity.Identity.DeviceScope.Contains(scopeId))
+                .ToList();
+
+            if (devices.Count() > 0)
+            {
+                // There shouldn't be more than one device, but even if there is
+                // for some reason, we can just use the first one
+                return Option.Some(devices.First());
+            }
+
+            return Option.None<ServiceIdentityTreeNode>();
+        }
+
+        /// <summary>
+        /// This is an object to wrap ServiceIdentity so that it can
+        /// point to a parent and act as a node in a tree.
+        /// </summary>
+        internal class ServiceIdentityTreeNode
+        {
+            public ServiceIdentity Identity { get; private set; }
+            public Option<ServiceIdentityTreeNode> Parent;
```

Comment:
>public Option<ServiceIdentityTreeNode> Parent; [](start = 12, length = 46)

From David: maybe carry a static string for the authchain in each node, so that we don't ever need to traverse upwards?

Reply:
Updates become tricky in that case. But it certainly seems efficient to keep a cache of the auth chains. Maybe in a different map that can be invalidated?